support-frontend/app/services/MParticleClient.scala

```diff
@@ -0,0 +1,169 @@
+    `type`: String,
+    value: String,
+)
+
+case class ProfileRequest(
+    environment_type: String,
+    identity: Identity,
+)
+
+case class AudienceMembership(
+    audience_id: Int,
+)
+
+case class ProfileResponse(
+    audience_memberships: List[AudienceMembership],
+)
+
+case class MParticleError(message: String) extends Throwable(message)
+
+object MParticleClient {
+  implicit val oauthTokenRequestEncoder: Encoder[OAuthTokenRequest] = deriveEncoder
+  implicit val oauthTokenResponseDecoder: Decoder[OAuthTokenResponse] = deriveDecoder
+  implicit val identityEncoder: Encoder[Identity] = deriveEncoder
+  implicit val profileRequestEncoder: Encoder[ProfileRequest] = deriveEncoder
+  implicit val audienceMembershipDecoder: Decoder[AudienceMembership] = deriveDecoder
+  implicit val profileResponseDecoder: Decoder[ProfileResponse] = deriveDecoder
+  implicit val mparticleErrorDecoder: Decoder[MParticleError] = deriveDecoder
+}
+
+class MParticleClient(
+    val httpClient: FutureHttpClient,
+    mparticleConfigProvider: MparticleConfigProvider,
+)(implicit ec: ExecutionContext)
+    extends WebServiceHelper[MParticleError]
+    with SafeLogging {
+
+  import MParticleClient._
+
+  private lazy val mparticleConfig: MparticleConfig = mparticleConfigProvider.get()
+
+  override val wsUrl: String = mparticleConfig.apiUrl
+  override val verboseLogging: Boolean = false
+
+  private case class CachedToken(token: String, expiresAt: DateTime)
+  private val tokenCache = new AtomicReference[Option[CachedToken]](None)
+  private val safetyMargin = 2.minutes
+
+  def getUserProfile(identityId: String): Future[MParticleUserProfile] = for {
+    accessToken <- getCachedAccessToken()
+    response <- callProfileAPI(identityId, accessToken.token)
+  } yield parseUserProfile(response)
+
+  private def getCachedAccessToken(): Future[MParticleAccessToken] = {
+    val now = DateTime.now()
+
+    tokenCache.get() match {
+      case Some(cachedToken) if cachedToken.expiresAt.isAfter(now) =>
+        Future.successful(MParticleAccessToken(cachedToken.token))
+      case _ =>
+        getAccessToken().map { case (token, expiresInSeconds) =>
+          val actualExpiryTime = now.plusSeconds(expiresInSeconds)
+          val safeExpiryTime = actualExpiryTime.minus(safetyMargin.toMillis)
+          tokenCache.set(Some(CachedToken(token, safeExpiryTime)))
+          logger.info(
+            s"Cached new mParticle access token, expires at: $safeExpiryTime (${safetyMargin.toSeconds}s safety margin)",
+          )
+          MParticleAccessToken(token)
+        }
+    }
+  }
+
+  private def getAccessToken(): Future[(String, Int)] = {
+    val request = OAuthTokenRequest(
+      client_id = mparticleConfig.clientId,
+      client_secret = mparticleConfig.clientSecret,
+      audience = mparticleConfig.apiUrl,
+      grant_type = "client_credentials",
+    )
+
+    // OAuth token endpoint is on a different domain, so we need to use a custom HTTP client call
```

Comment:
did you consider making a separate MParticleAuthClient extends WebServiceHelper?  It doesn't have to be in a separate file - it can be a nested class within this one, or a separate one below in the same file